Issue 2967/filter events by type

felixbuchholz opened the pull request:
## Description
This PR introduces a event type filter for events on these pages
My page: http://localhost:3000/my/feed
Organization page: http://localhost:3000/o/1
Project page: http://localhost:3000/o/1/projects/168

## Screenshots
<img width="377" height="674" alt="screen 2025-08-31 at 17 35 17" src="https://github.com/user-attachments/assets/517266ee-31f1-4185-b743-0ca6403ee9f1" />

<img width="392" height="845" alt="screen 2025-08-31 at 17 36 58" src="https://github.com/user-attachments/assets/68d291ea-80a8-4a5c-bbf7-bb9b8d0bfb38" />

Note that we are cutting of the filter buttons and horizontal scroll is necessary. Not sure if we want to do something about that.


## Changes
- Add a new `useEventTypeFilter` hook that does not keep track of the filter state itself but allows for interop between local and redux state management.
   - Wanted to keep it a bit more reusable. But if the abstraction is not useful, we can drop the last commit and only add the functionality based on the patterns for the other filters manually without any sharing code.
- No changes on the existing filters


## Notes to reviewer
- Check the three different event pages (replace localhost with the preview deploy link)
- Test for a event type with the same title across multiple orgs
- Validate clearing the filters work ('X' button and 'Clear filters')


## Related issues
Resolves #2967

richardolsson reviewed (changes requested):
Awesome work! This will be greatly appreciated by many users.

I haven't had time to go through all the code yet, but I wanted to play around with the preview build and I really like it. The only problem I found is that it renders a zero in some cases. Details below. 👇

richardolsson reviewed (commented):
Nice work! I really wanted to merge this, but I do think there are some things that we should figure out. Can you have a look and see what you think?

richardolsson commented on src/features/organizations/pages/PublicOrgPage.tsx:
```diff
@@ -72,9 +80,10 @@ const PublicOrgPage: FC<Props> = ({ orgId }) => {
+    eventTypeFilter.isFiltered ||
     !!geojsonToFilterBy.length ||
-    !!dateFilterState;
+    orgIdsToFilterBy.length;
```
Very fast, and confirmed working. 👏

felixbuchholz commented on src/features/events/hooks/useEventTypeFilter.ts:
```diff
@@ -0,0 +1,76 @@
+  }
+) => {
+  const eventTypes = useMemo(() => {
+    const uniqueTypes = new Set(events.map(getEventTypeFromEvent));
+    return Array.from(uniqueTypes).sort((a, b) => {
```
Sounds good to me. I'd want to show the event type options as `Sentence case` then.

felixbuchholz replied:
Small update in this commit https://github.com/zetkin/app.zetkin.org/pull/3012/commits/2a04ae04eb060f855a3aed6589cd057a24f875be

felixbuchholz replied:
Good point!
After my initial demo I saw that `EventTypeAutocomplete` is supposed to enforce sentence case. I just refactored it and aligned it in the filters. I don't remember how I was able to create an all caps event type before. If we want to allow other cases I will also go ahead and remove that for the autocomplete component, right?

Here's a quick demo

https://github.com/user-attachments/assets/5bb8ceed-a0e4-4976-bc68-d7638b45029d

And the code in main before my PR

https://github.com/zetkin/app.zetkin.org/blob/3ad46c895d724f18ebbe0c85e5360b29503c297b/src/features/events/components/EventTypeAutocomplete.tsx#L147-L152

felixbuchholz commented on src/features/campaigns/pages/PublicProjPage.tsx:
```diff
@@ -363,6 +400,31 @@ const PublicProjectPage: FC<Props> = ({ campId, orgId }) => {
           />
         </Box>
       </ZUIDrawerModal>
+      <ZUIDrawerModal
+        onClose={() => setDrawerContent(null)}
+        open={drawerContent == 'eventTypes'}
+      >
+        <List>
+          {eventTypeFilter.eventTypes.map((eventType) => (
+            <ListItem
+              key={eventTypeFilter.getLabelFromEventType(eventType)}
+              sx={{ justifyContent: 'space-between' }}
+            >
+              <Box alignItems="center" display="flex">
+                <ZUIText>
+                  {eventTypeFilter.getLabelFromEventType(eventType)}
+                </ZUIText>
+              </Box>
+              <Switch
+                checked={eventTypesToFilterBy.includes(eventType)}
+                onChange={() => {
+                  eventTypeFilter.toggleEventType(eventType);
+                }}
+              />
+            </ListItem>
+          ))}
+        </List>
+      </ZUIDrawerModal>
```
I considered that, but ended up being ok with how clear the composition of the current version is. The component boundary as you suggested would require quite a few props and good names :)
I can take a stab at it.

felixbuchholz replied:
Sorry, running out of time and I can't promise I will get to this before after the next hackathon. Maybe in a separate PR?

felixbuchholz replied:
Started a discussion in the code-red slack channel to think more broader about how to make event filters more reusable and convenient. Will follow up on this based on the feedback there. :)
[Link](https://zetkin.slack.com/archives/C063B88MLMS/p1758884815261489)

felixbuchholz commented on src/features/campaigns/l10n/messageIds.ts:
```diff
@@ -167,6 +167,9 @@ export default makeMessages('feat.campaigns', {
         removeFiltersButton: m('Clear filters'),
       },
       filterButtonLabels: {
+        eventTypes: m<{ numEventTypes: number }>(
```
Moved the label logic into the hook and defined the message in the core events. Not sure we've used this pattern before, but seems convenient to have the label available everywhere the hook is used. Let me know what you think!

https://github.com/zetkin/app.zetkin.org/pull/3012/commits/6565fbd5d4a686fd453ae58cbe97a6b7fb536be9

felixbuchholz commented on src/features/campaigns/l10n/messageIds.ts:
```diff
@@ -167,6 +167,9 @@ export default makeMessages('feat.campaigns', {
       },
       filterButtonLabels: {
+        eventTypes: m<{ numEventTypes: number }>(
+          '{numEventTypes, plural,=0 {Event types} =1 {1 event type} other {# event types}}'
```
Updated in https://github.com/zetkin/app.zetkin.org/pull/3012/commits/9c5e3b2711e65b0ea2c53bb0fde84a3794c7634a

<img width="1152" height="1442" alt="image" src="https://github.com/user-attachments/assets/d632c12f-c834-4179-8b13-7de1d20339a8" />

felixbuchholz commented:
@richardolsson addressed your comments except for one. I think the PR improved quite a bit.
I can try small fixes if necessary, but won't be fully back for bigger updates until Sep 19th.

richardolsson commented on src/features/organizations/pages/PublicOrgPage.tsx:
```diff
@@ -477,6 +500,31 @@ const PublicOrgPage: FC<Props> = ({ orgId }) => {
           ))}
         </List>
       </ZUIDrawerModal>
+      <ZUIDrawerModal
+        onClose={() => setDrawerContent(null)}
+        open={drawerContent == 'eventTypes'}
+      >
```
Confirmed to work. Fixes #3052

richardolsson reviewed (approved):
Really good work on this! I'm sorry for the delay in reviewing it, but I don't want to hold it back any longer, so lets merge.

There are some minor things, both that you have mentioned and that I have commented on in my previous PR. I'll leave those comments unresolved above. If you want to create a separate PR that would be great, but no rush.

felixbuchholz commented:
Thanks, I'll keep the small things on my list!
